## Data Collection & Preparation

* ``Pandas`` for managing the data
* ``NumPy`` for math operations
* ``Sklearn`` for Machine Learning
* ``Seaborn`` for advanced visualization
* ``Matplotlib`` for additional plotting

In [1]:
# Import libraries 

import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_squared_error, r2_score
import warnings
from datetime import datetime, timedelta
warnings.filterwarnings("ignore", category=UserWarning)
%matplotlib inline

In [2]:
#read data files

cs = pd.read_csv("charging_sessions.csv")
wba = pd.read_csv("weather_burbank_airport.csv")

In [3]:
#print the first 5 rows of the charging_sessions dataset
cs.head()

Unnamed: 0                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...

In [4]:
#print out the first 5 rows of the weather burbank airport dataset
wba.head()

city            timestamp  temperature  cloud_cover  \
0  Burbank  2018-01-01 08:53:00          9.0         33.0   
1  Burbank  2018-01-01 09:53:00          9.0         33.0   
2  Burbank  2018-01-01 10:53:00          9.0         21.0   
3  Burbank  2018-01-01 11:53:00          9.0         29.0   
4  Burbank  2018-01-01 12:53:00          8.0         33.0   

  cloud_cover_description  pressure  windspeed  precipitation  \
0                    Fair    991.75        9.0            0.0   
1                    Fair    992.08        0.0            0.0   
2                    Haze    992.08        0.0            0.0   
3           Partly Cloudy    992.08        0.0            0.0   
4                    Fair    992.08        0.0            0.0   

   felt_temperature  
0               8.0  
1               9.0  
2               9.0  
3               9.0  
4               8.0

In [5]:
#print out the columns of charging sessions dataset
cs.columns

Index(['Unnamed: 0', 'id', 'connectionTime', 'disconnectTime',
       'doneChargingTime', 'kWhDelivered', 'sessionID', 'siteID', 'spaceID',
       'stationID', 'timezone', 'userID', 'userInputs'],
      dtype='object')

In [6]:
#print out the columns of weather burbank airport dataset
wba.columns

Index(['city', 'timestamp', 'temperature', 'cloud_cover',
       'cloud_cover_description', 'pressure', 'windspeed', 'precipitation',
       'felt_temperature'],
      dtype='object')

In [7]:
#print out the tail of cs dataset to see how many entries there are
cs.tail()

Unnamed: 0                        id             connectionTime  \
66445       10083  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446       10084  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447       10085  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448       10086  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   
66449       10087  5d574ad2f9af8b4c10c03656  2019-07-31 19:21:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   
66449  2019-07-31 22:00:04+00:00  2019-07-31 20:51:34+00:00         0.897   

                                    sessionID  siteID  spaceID    stationID  \
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   
66449  1_1_178_817_2019-07-31 19:21:46.727697       1  AG-1F09  1-1-178-817   

                  timezone  userID  \
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   
66449  America/Los_Angeles     NaN   

                                              userInputs  
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...  
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...  
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...  
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...  
66449                                                NaN

In [8]:
#print out the tail of cs dataset to see how many entries there are
wba.tail()

city            timestamp  temperature  cloud_cover  \
29239  Burbank  2021-01-01 03:53:00         13.0         33.0   
29240  Burbank  2021-01-01 04:53:00         12.0         33.0   
29241  Burbank  2021-01-01 05:53:00         12.0         33.0   
29242  Burbank  2021-01-01 06:53:00         11.0         33.0   
29243  Burbank  2021-01-01 07:53:00         10.0         33.0   

      cloud_cover_description  pressure  windspeed  precipitation  \
29239                    Fair    986.81        0.0            0.0   
29240                    Fair    986.81       11.0            0.0   
29241                    Fair    987.47        9.0            0.0   
29242                    Fair    987.14       13.0            0.0   
29243                    Fair    987.80        6.0            0.0   

       felt_temperature  
29239              13.0  
29240              12.0  
29241              12.0  
29242              11.0  
29243              10.0

### We will focus on the charging_sessions dataset for a while

In [9]:
#print information about cs dataset
cs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


We can see that this dataset has an unnamed column that represents the entry number. For better readability, we will rename this column

In [10]:
#rename the unnamed column
cs.rename( columns={'Unnamed: 0':'number'}, inplace=True )

In [11]:
#check if the column was renamed
cs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   number            66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


From looking at the count of every column, we can see that the "doneChargingTime" is missing values. The missing values in the "userID" and "userInputs" columns are because the user was not registered, so we can ignore them for now.

**Detecting missing numerical data**

In [12]:
#check if we can spot null-values
cs.isnull()

number     id  connectionTime  disconnectTime  doneChargingTime  \
0       False  False           False           False             False   
1       False  False           False           False             False   
2       False  False           False           False             False   
3       False  False           False           False             False   
4       False  False           False           False             False   
...       ...    ...             ...             ...               ...   
66445   False  False           False           False             False   
66446   False  False           False           False             False   
66447   False  False           False           False             False   
66448   False  False           False           False             False   
66449   False  False           False           False             False   

       kWhDelivered  sessionID  siteID  spaceID  stationID  timezone  userID  \
0             False      False   False    False      False     False   False   
1             False      False   False    False      False     False   False   
2             False      False   False    False      False     False   False   
3             False      False   False    False      False     False   False   
4             False      False   False    False      False     False   False   
...             ...        ...     ...      ...        ...       ...     ...   
66445         False      False   False    False      False     False   False   
66446         False      False   False    False      False     False   False   
66447         False      False   False    False      False     False   False   
66448         False      False   False    False      False     False   False   
66449         False      False   False    False      False     False    True   

       userInputs  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
66445       False  
66446       False  
66447       False  
66448       False  
66449        True  

[66450 rows x 13 columns]

In [13]:
#drop all rows containing null values
cs.dropna(axis=0, inplace=False)

number                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
...       ...                       ...                        ...   
66444   10082  5d574ad2f9af8b4c10c03651  2019-07-31 14:50:17+00:00   
66445   10083  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446   10084  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447   10085  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448   10086  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
0      2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1      2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2      2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3      2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4      2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
...                          ...                        ...           ...   
66444  2019-08-01 01:01:18+00:00  2019-07-31 20:29:24+00:00        27.174   
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
66444  1_1_191_784_2019-07-31 14:50:17.037367       1  AG-4F40  1-1-191-784   
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   

                  timezone  userID  \
0      America/Los_Angeles   194.0   
1      America/Los_Angeles  4275.0   
2      America/Los_Angeles   344.0   
3      America/Los_Angeles  1117.0   
4      America/Los_Angeles   334.0   
...                    ...     ...   
66444  America/Los_Angeles   448.0   
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   

                                              userInputs  
0      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1      [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4      [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...  
...                                                  ...  
66444  [{'WhPerMile': 200, 'kWhRequested': 28.0, 'mil...  
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...  
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...  
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...  
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...  

[46008 rows x 13 columns]

In [14]:
#check how many rows are getting dropped
len(cs)-len(cs.dropna())

20442

This number is higher than expected, but because this factors in the null values from userID and userInputs(unregistered users or users with no Input), we want so specifically look for missing values in the "doneChargingTime"-column

In [15]:
len(cs)-len(cs.dropna(subset = ['doneChargingTime']))

4088

This number tells us that there are 4088 sessions with no timestamp for when the EV got done charging, meaning that it didn't fully charge. Let's look at this data.

In [16]:
#display records with 'doneCharging = Null
cs_notdone = cs[cs['doneChargingTime'].isnull()]
cs_notdone

number                        id             connectionTime  \
3756     3756  5ea3848df9af8b428bf97225  2020-04-08 16:36:53+00:00   
3757     3757  5ea4d60df9af8b46573f1ee7  2020-04-09 13:23:52+00:00   
3758     3758  5ea4d60df9af8b46573f1ee8  2020-04-09 13:40:03+00:00   
3759     3759  5ea4d60df9af8b46573f1ee9  2020-04-09 14:01:53+00:00   
3760     3760  5ea4d60df9af8b46573f1eea  2020-04-09 15:01:06+00:00   
...       ...                       ...                        ...   
60411    4049  5cb3d2a8f9af8b4551261e75  2019-03-30 00:32:10+00:00   
61097    4735  5cc8e891f9af8b45d975b819  2019-04-14 16:19:12+00:00   
61103    4741  5cca3a54f9af8b49aaa4cbac  2019-04-15 13:00:59+00:00   
61117    4755  5cca3a54f9af8b49aaa4cbba  2019-04-15 13:52:45+00:00   
65530    9168  5d3f8f96f9af8b7f97666a55  2019-07-13 20:13:18+00:00   

                  disconnectTime doneChargingTime  kWhDelivered  \
3756   2020-04-09 03:43:29+00:00              NaN        25.183   
3757   2020-04-09 17:59:02+00:00              NaN        14.136   
3758   2020-04-09 21:43:22+00:00              NaN        14.203   
3759   2020-04-09 16:33:40+00:00              NaN        13.041   
3760   2020-04-09 23:38:08+00:00              NaN         5.797   
...                          ...              ...           ...   
60411  2019-03-30 02:23:14+00:00              NaN        11.148   
61097  2019-04-14 20:38:39+00:00              NaN        11.678   
61103  2019-04-15 13:55:25+00:00              NaN         3.343   
61117  2019-04-15 14:47:01+00:00              NaN         4.558   
65530  2019-07-13 20:37:40+00:00              NaN         2.218   

                                    sessionID  siteID  spaceID    stationID  \
3756   1_1_194_822_2020-04-08 16:36:53.316264       1  AG-1F12  1-1-194-822   
3757   1_1_193_827_2020-04-09 13:23:51.689520       1  AG-1F02  1-1-193-827   
3758   1_1_178_823_2020-04-09 13:39:55.115497       1  AG-1F08  1-1-178-823   
3759   1_1_178_828_2020-04-09 14:01:53.467693       1  AG-1F10  1-1-178-828   
3760   1_1_193_820_2020-04-09 15:01:06.343454       1  AG-1F04  1-1-193-820   
...                                       ...     ...      ...          ...   
60411  1_1_193_829_2019-03-30 00:32:10.128651       1  AG-1F03  1-1-193-829   
61097  1_1_193_816_2019-04-14 16:19:11.864262       1  AG-1F05  1-1-193-816   
61103  1_1_193_819_2019-04-15 13:00:58.722284       1  AG-1F06  1-1-193-819   
61117  1_1_179_800_2019-04-15 13:52:44.693153       1  AG-3F32  1-1-179-800   
65530  1_1_193_825_2019-07-13 20:13:18.490314       1  AG-1F01  1-1-193-825   

                  timezone  userID  \
3756   America/Los_Angeles   507.0   
3757   America/Los_Angeles   419.0   
3758   America/Los_Angeles  5065.0   
3759   America/Los_Angeles   651.0   
3760   America/Los_Angeles   466.0   
...                    ...     ...   
60411  America/Los_Angeles   194.0   
61097  America/Los_Angeles   364.0   
61103  America/Los_Angeles   651.0   
61117  America/Los_Angeles   194.0   
65530  America/Los_Angeles    76.0   

                                              userInputs  
3756   [{'WhPerMile': 250, 'kWhRequested': 75.0, 'mil...  
3757   [{'WhPerMile': 400, 'kWhRequested': 40.0, 'mil...  
3758   [{'WhPerMile': 400, 'kWhRequested': 24.0, 'mil...  
3759   [{'WhPerMile': 400, 'kWhRequested': 20.0, 'mil...  
3760   [{'WhPerMile': 300, 'kWhRequested': 6.0, 'mile...  
...                                                  ...  
60411  [{'WhPerMile': 400, 'kWhRequested': 24.0, 'mil...  
61097  [{'WhPerMile': 400, 'kWhRequested': 56.0, 'mil...  
61103  [{'WhPerMile': 400, 'kWhRequested': 20.0, 'mil...  
61117  [{'WhPerMile': 400, 'kWhRequested': 12.0, 'mil...  
65530  [{'WhPerMile': 301, 'kWhRequested': 9.03, 'mil...  

[4088 rows x 13 columns]

### Dealing with missing values

When dealing with missing values, we can either **eliminate** them from the dataset or **impute** the null values with estimates. Because the missing data points are timestamps, it's hard them because they simply do not exist. The EV never got done charging, so estimating a value doesn't make any sense. And because our job is to optimize utilization, looking at data points where the EV never fully charged(and thus didn't stay plugged in without charging and taking up charging space) is not adequate. That is why we will drop these data points.

In [17]:
#we take every row where "doneChargingTime" is not null into our new dataset cs_clean
cs_clean = cs[cs['doneChargingTime'].notna()]

In [18]:
cs_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 62362 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   number            62362 non-null  int64  
 1   id                62362 non-null  object 
 2   connectionTime    62362 non-null  object 
 3   disconnectTime    62362 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      62362 non-null  float64
 6   sessionID         62362 non-null  object 
 7   siteID            62362 non-null  int64  
 8   spaceID           62362 non-null  object 
 9   stationID         62362 non-null  object 
 10  timezone          62362 non-null  object 
 11  userID            46008 non-null  float64
 12  userInputs        46008 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.7+ MB


The data is now clean and we can work with it!

# Data Wrangling & Feature Engineering

To better understand what typical charging sessions look like, we will do a cluster analysis. But before that, we need to do some more Data Wrangling and Feature Engineering.

Lets do some more data wrangling and Feature Engineering. We already identified that site1 is the private site and site2 is the public site.

In [19]:
#From Descriptive Analytics

dummy_var1 = pd.get_dummies(cs_clean["siteID"])
dummy_var1.head()
dummy_var1.rename(columns={1:'private-site', 2:'public-site'}, inplace=True)
dummy_var1.head()
cs_clean = pd.concat([cs_clean, dummy_var1], axis=1)
cs_clean.drop("siteID", axis = 1, inplace=True)

cs_clean.head()

number                        id             connectionTime  \
0       0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1       1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2       2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3       3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4       4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   

                                sessionID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  private-site  \
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...          True   
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...          True   
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...          True   

   public-site  
0        False  
1        False  
2        False  
3        False  
4        False

In [20]:
cs_clean.dtypes

number                int64
id                   object
connectionTime       object
disconnectTime       object
doneChargingTime     object
kWhDelivered        float64
sessionID            object
spaceID              object
stationID            object
timezone             object
userID              float64
userInputs           object
private-site           bool
public-site            bool
dtype: object

To be able to work with Time data, we first have to convert all time data to datetime64. Then, we have to convert the times to match the Los Angeles Timezone.

In [21]:
cs_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 62362 entries, 0 to 66449
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   number            62362 non-null  int64  
 1   id                62362 non-null  object 
 2   connectionTime    62362 non-null  object 
 3   disconnectTime    62362 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      62362 non-null  float64
 6   sessionID         62362 non-null  object 
 7   spaceID           62362 non-null  object 
 8   stationID         62362 non-null  object 
 9   timezone          62362 non-null  object 
 10  userID            46008 non-null  float64
 11  userInputs        46008 non-null  object 
 12  private-site      62362 non-null  bool   
 13  public-site       62362 non-null  bool   
dtypes: bool(2), float64(2), int64(1), object(9)
memory usage: 6.3+ MB


We see that datetimes are for UTC(UTC+0) instead of UTC-8, so we have to correct that

In [22]:
cs_clean['connectionTime'] = pd.to_datetime(cs_clean['connectionTime']) - pd.Timedelta(hours=8)
cs_clean['disconnectTime'] = pd.to_datetime(cs_clean['disconnectTime']) - pd.Timedelta(hours=8)
cs_clean['doneChargingTime'] = pd.to_datetime(cs_clean['doneChargingTime']) - pd.Timedelta(hours=8)
cs_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 62362 entries, 0 to 66449
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype              
---  ------            --------------  -----              
 0   number            62362 non-null  int64              
 1   id                62362 non-null  object             
 2   connectionTime    62362 non-null  datetime64[ns, UTC]
 3   disconnectTime    62362 non-null  datetime64[ns, UTC]
 4   doneChargingTime  62362 non-null  datetime64[ns, UTC]
 5   kWhDelivered      62362 non-null  float64            
 6   sessionID         62362 non-null  object             
 7   spaceID           62362 non-null  object             
 8   stationID         62362 non-null  object             
 9   timezone          62362 non-null  object             
 10  userID            46008 non-null  float64            
 11  userInputs        46008 non-null  object             
 12  private-site      62362 non-null  bool               
 13  public

In [23]:
cs_clean.head()

number                        id            connectionTime  \
0       0  5e23b149f9af8b5fe4b973cf 2020-01-02 05:08:54+00:00   
1       1  5e23b149f9af8b5fe4b973d0 2020-01-02 05:36:50+00:00   
2       2  5e23b149f9af8b5fe4b973d1 2020-01-02 05:56:35+00:00   
3       3  5e23b149f9af8b5fe4b973d2 2020-01-02 05:59:58+00:00   
4       4  5e23b149f9af8b5fe4b973d3 2020-01-02 06:00:01+00:00   

             disconnectTime          doneChargingTime  kWhDelivered  \
0 2020-01-02 11:11:15+00:00 2020-01-02 09:31:35+00:00        25.016   
1 2020-01-02 14:38:21+00:00 2020-01-02 12:18:05+00:00        33.097   
2 2020-01-02 16:39:22+00:00 2020-01-02 08:35:06+00:00         6.521   
3 2020-01-02 08:38:39+00:00 2020-01-02 07:18:45+00:00         2.355   
4 2020-01-02 14:08:40+00:00 2020-01-02 10:17:30+00:00        13.375   

                                sessionID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  private-site  \
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...          True   
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...          True   
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...          True   

   public-site  
0        False  
1        False  
2        False  
3        False  
4        False

We want to see what day of the week the session took place in, so lets compute a new column "dayOfWeek".

In [24]:
cs_clean['dayOfWeek'] = cs_clean['connectionTime'].dt.day_name()

cs_clean.head()

number                        id            connectionTime  \
0       0  5e23b149f9af8b5fe4b973cf 2020-01-02 05:08:54+00:00   
1       1  5e23b149f9af8b5fe4b973d0 2020-01-02 05:36:50+00:00   
2       2  5e23b149f9af8b5fe4b973d1 2020-01-02 05:56:35+00:00   
3       3  5e23b149f9af8b5fe4b973d2 2020-01-02 05:59:58+00:00   
4       4  5e23b149f9af8b5fe4b973d3 2020-01-02 06:00:01+00:00   

             disconnectTime          doneChargingTime  kWhDelivered  \
0 2020-01-02 11:11:15+00:00 2020-01-02 09:31:35+00:00        25.016   
1 2020-01-02 14:38:21+00:00 2020-01-02 12:18:05+00:00        33.097   
2 2020-01-02 16:39:22+00:00 2020-01-02 08:35:06+00:00         6.521   
3 2020-01-02 08:38:39+00:00 2020-01-02 07:18:45+00:00         2.355   
4 2020-01-02 14:08:40+00:00 2020-01-02 10:17:30+00:00        13.375   

                                sessionID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  private-site  \
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...          True   
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...          True   
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...          True   

   public-site dayOfWeek  
0        False  Thursday  
1        False  Thursday  
2        False  Thursday  
3        False  Thursday  
4        False  Thursday

We now want to compute some more columns that might give us a better understanding of the data. Lets start with chargeduration. Chargeduration tells us for how long the session went on.

In [25]:
cs_clean['chargeDuration'] = cs_clean['disconnectTime'] - cs_clean['connectionTime']
cs_clean['chargeDurationHours'] = (cs_clean['chargeDuration'] / pd.Timedelta(minutes=1)).astype(float)/60

cs_clean.head()

number                        id            connectionTime  \
0       0  5e23b149f9af8b5fe4b973cf 2020-01-02 05:08:54+00:00   
1       1  5e23b149f9af8b5fe4b973d0 2020-01-02 05:36:50+00:00   
2       2  5e23b149f9af8b5fe4b973d1 2020-01-02 05:56:35+00:00   
3       3  5e23b149f9af8b5fe4b973d2 2020-01-02 05:59:58+00:00   
4       4  5e23b149f9af8b5fe4b973d3 2020-01-02 06:00:01+00:00   

             disconnectTime          doneChargingTime  kWhDelivered  \
0 2020-01-02 11:11:15+00:00 2020-01-02 09:31:35+00:00        25.016   
1 2020-01-02 14:38:21+00:00 2020-01-02 12:18:05+00:00        33.097   
2 2020-01-02 16:39:22+00:00 2020-01-02 08:35:06+00:00         6.521   
3 2020-01-02 08:38:39+00:00 2020-01-02 07:18:45+00:00         2.355   
4 2020-01-02 14:08:40+00:00 2020-01-02 10:17:30+00:00        13.375   

                                sessionID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  private-site  \
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...          True   
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...          True   
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...          True   

   public-site dayOfWeek  chargeDuration  chargeDurationHours  
0        False  Thursday 0 days 06:02:21             6.039167  
1        False  Thursday 0 days 09:01:31             9.025278  
2        False  Thursday 0 days 10:42:47            10.713056  
3        False  Thursday 0 days 02:38:41             2.644722  
4        False  Thursday 0 days 08:08:39             8.144167

We want another colums that displays the excess utilization(idle) time, that the car utilized the station when it was already fully loaded.

In [26]:
cs_clean['excessTime'] = cs_clean['disconnectTime'] - cs_clean['doneChargingTime']
cs_clean['excessDurationHours'] = (cs_clean['excessTime'] / pd.Timedelta(minutes=1)).astype(float)/60

cs_clean

number                        id            connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf 2020-01-02 05:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0 2020-01-02 05:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1 2020-01-02 05:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2 2020-01-02 05:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3 2020-01-02 06:00:01+00:00   
...       ...                       ...                       ...   
66445   10083  5d574ad2f9af8b4c10c03652 2019-07-31 10:08:04+00:00   
66446   10084  5d574ad2f9af8b4c10c03653 2019-07-31 10:40:41+00:00   
66447   10085  5d574ad2f9af8b4c10c03654 2019-07-31 11:04:40+00:00   
66448   10086  5d574ad2f9af8b4c10c03655 2019-07-31 11:19:47+00:00   
66449   10087  5d574ad2f9af8b4c10c03656 2019-07-31 11:21:47+00:00   

                 disconnectTime          doneChargingTime  kWhDelivered  \
0     2020-01-02 11:11:15+00:00 2020-01-02 09:31:35+00:00        25.016   
1     2020-01-02 14:38:21+00:00 2020-01-02 12:18:05+00:00        33.097   
2     2020-01-02 16:39:22+00:00 2020-01-02 08:35:06+00:00         6.521   
3     2020-01-02 08:38:39+00:00 2020-01-02 07:18:45+00:00         2.355   
4     2020-01-02 14:08:40+00:00 2020-01-02 10:17:30+00:00        13.375   
...                         ...                       ...           ...   
66445 2019-07-31 15:29:18+00:00 2019-07-31 15:30:18+00:00        28.787   
66446 2019-07-31 16:59:42+00:00 2019-07-31 13:44:23+00:00         7.787   
66447 2019-07-31 14:44:22+00:00 2019-07-31 14:45:21+00:00        11.274   
66448 2019-07-31 16:34:51+00:00 2019-07-31 13:25:30+00:00        11.589   
66449 2019-07-31 14:00:04+00:00 2019-07-31 12:51:34+00:00         0.897   

                                    sessionID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967  AG-1F06  1-1-193-819   
...                                       ...      ...          ...   
66445  1_1_179_809_2019-07-31 18:08:04.432654  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358  AG-4F43  1-1-191-778   
66449  1_1_178_817_2019-07-31 19:21:46.727697  AG-1F09  1-1-178-817   

                  timezone  userID  \
0      America/Los_Angeles   194.0   
1      America/Los_Angeles  4275.0   
2      America/Los_Angeles   344.0   
3      America/Los_Angeles  1117.0   
4      America/Los_Angeles   334.0   
...                    ...     ...   
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   
66449  America/Los_Angeles     NaN   

                                              userInputs  private-site  \
0      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...          True   
1      [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...          True   
2      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
3      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
4      [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...          True   
...                                                  ...           ...   
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...          True   
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...          True   
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...          True   
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...          True   
66449                                                NaN          True   

       public-site  dayOfWeek  chargeDuration  chargeDurationHours  \
0       

Lets make a column that displays the **efficiency** of the session. We calculate the efficiency of the session by dividing the excessTime by the chargeDuration. A 100% efficiency means that the EV disconnected right when it was fully loaded(0 idle time).

In [27]:
cs_clean['sessionEfficiency'] = (1-cs_clean['excessTime'] / cs_clean['chargeDuration'])*100

cs_clean

number                        id            connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf 2020-01-02 05:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0 2020-01-02 05:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1 2020-01-02 05:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2 2020-01-02 05:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3 2020-01-02 06:00:01+00:00   
...       ...                       ...                       ...   
66445   10083  5d574ad2f9af8b4c10c03652 2019-07-31 10:08:04+00:00   
66446   10084  5d574ad2f9af8b4c10c03653 2019-07-31 10:40:41+00:00   
66447   10085  5d574ad2f9af8b4c10c03654 2019-07-31 11:04:40+00:00   
66448   10086  5d574ad2f9af8b4c10c03655 2019-07-31 11:19:47+00:00   
66449   10087  5d574ad2f9af8b4c10c03656 2019-07-31 11:21:47+00:00   

                 disconnectTime          doneChargingTime  kWhDelivered  \
0     2020-01-02 11:11:15+00:00 2020-01-02 09:31:35+00:00        25.016   
1     2020-01-02 14:38:21+00:00 2020-01-02 12:18:05+00:00        33.097   
2     2020-01-02 16:39:22+00:00 2020-01-02 08:35:06+00:00         6.521   
3     2020-01-02 08:38:39+00:00 2020-01-02 07:18:45+00:00         2.355   
4     2020-01-02 14:08:40+00:00 2020-01-02 10:17:30+00:00        13.375   
...                         ...                       ...           ...   
66445 2019-07-31 15:29:18+00:00 2019-07-31 15:30:18+00:00        28.787   
66446 2019-07-31 16:59:42+00:00 2019-07-31 13:44:23+00:00         7.787   
66447 2019-07-31 14:44:22+00:00 2019-07-31 14:45:21+00:00        11.274   
66448 2019-07-31 16:34:51+00:00 2019-07-31 13:25:30+00:00        11.589   
66449 2019-07-31 14:00:04+00:00 2019-07-31 12:51:34+00:00         0.897   

                                    sessionID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967  AG-1F06  1-1-193-819   
...                                       ...      ...          ...   
66445  1_1_179_809_2019-07-31 18:08:04.432654  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358  AG-4F43  1-1-191-778   
66449  1_1_178_817_2019-07-31 19:21:46.727697  AG-1F09  1-1-178-817   

                  timezone  userID  \
0      America/Los_Angeles   194.0   
1      America/Los_Angeles  4275.0   
2      America/Los_Angeles   344.0   
3      America/Los_Angeles  1117.0   
4      America/Los_Angeles   334.0   
...                    ...     ...   
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   
66449  America/Los_Angeles     NaN   

                                              userInputs  private-site  \
0      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...          True   
1      [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...          True   
2      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
3      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...          True   
4      [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...          True   
...                                                  ...           ...   
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...          True   
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...          True   
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...          True   
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...          True   
66449                                                NaN          True   

       public-site  dayOfWeek  chargeDuration  chargeDurationHours  \
0       

As we can see below, there are 27 entries in the data set where the doneCharginTime is BEFORE the connectionTime. Since that does not make sense, we'll remove them from the data.

In [28]:
out_of_range_entries = cs_clean[cs_clean['connectionTime'] > cs_clean['doneChargingTime']]
len(out_of_range_entries)

27

In [29]:
cs_clean = cs_clean[cs_clean['connectionTime'] < cs_clean['doneChargingTime']]

We now want to see when the session took place throughout the day. We can achieve this by creating new columns named connectionHour and disconnectHour. These columns show the decimals of the hours when a session started and when it ended.

In [30]:
cs_clean['connectionHour'] = cs_clean['connectionTime'].dt.hour + cs_clean['connectionTime'].dt.minute / 60
cs_clean['disconnectHour'] = cs_clean['disconnectTime'].dt.hour + cs_clean['disconnectTime'].dt.minute / 60

C:\Users\Timo\AppData\Local\Temp\ipykernel_20876\1603379457.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cs_clean['connectionHour'] = cs_clean['connectionTime'].dt.hour + cs_clean['connectionTime'].dt.minute / 60
C:\Users\Timo\AppData\Local\Temp\ipykernel_20876\1603379457.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cs_clean['disconnectHour'] = cs_clean['disconnectTime'].dt.hour + cs_clean['disconnectTime'].dt.minute / 60


To prepare the forecast we are trying to create, we need a DataFrame that aggregates the number of users for each hour per site. We will create these two DataFrames, for the private and public site respectively, step by step.

To make this easier, we will just include the total number of users within each hour. That means if a user started charging at 05:59:00, he will be included in the data for 05:00:00. If the session this user started ended at 09:16:00, the session will also be added to the hours 06:00:00, 07:00:00, 08:00:00, and 09:00:00.

In [78]:
# First, we create a new DataFrame for the private site that includes each hour within each charging session
timeseries_private = pd.DataFrame()

# We iterate over each row in cs_clean for private-site
for _, row in cs_clean[cs_clean['private-site']].iterrows():
    # We generate the range of hours that the charging session took place in
    hours_range = pd.date_range(start=row['connectionTime'], end=row['disconnectTime'], freq='H')
    
    # We need to truncate the timestamps to only include the hour part
    hours_range = hours_range.floor('H')
    
    # We need to add each hour to the new DataFrame
    timeseries_private = pd.concat([timeseries_private, pd.DataFrame({'Hour': hours_range})], ignore_index=True)

# Finally, we count the number of users for each unique hour
timeseries_private = timeseries_private.groupby(['Hour']).size().reset_index(name='numberOfUsers_private')

# We create a separate DataFrame for public-site with the exact same steps
timeseries_public = pd.DataFrame()

for _, row in cs_clean[cs_clean['public-site']].iterrows():
    hours_range = pd.date_range(start=row['connectionTime'], end=row['disconnectTime'], freq='H')
    hours_range = hours_range.floor('H')
    timeseries_public = pd.concat([timeseries_public, pd.DataFrame({'Hour': hours_range})], ignore_index=True)

timeseries_public = timeseries_public.groupby(['Hour']).size().reset_index(name='numberOfUsers_public')

Now, we take a look at the DataFrame for the private site that we have just created.

In [79]:
timeseries_private.head(25)

Hour  numberOfUsers_private
0  2018-10-08 05:00:00+00:00                      4
1  2018-10-08 06:00:00+00:00                     19
2  2018-10-08 07:00:00+00:00                     19
3  2018-10-08 08:00:00+00:00                     19
4  2018-10-08 09:00:00+00:00                     19
5  2018-10-08 10:00:00+00:00                     22
6  2018-10-08 11:00:00+00:00                     24
7  2018-10-08 12:00:00+00:00                     26
8  2018-10-08 13:00:00+00:00                     26
9  2018-10-08 14:00:00+00:00                     30
10 2018-10-08 15:00:00+00:00                     34
11 2018-10-08 16:00:00+00:00                     25
12 2018-10-08 17:00:00+00:00                      4
13 2018-10-08 18:00:00+00:00                      3
14 2018-10-08 19:00:00+00:00                      1
15 2018-10-08 20:00:00+00:00                      1
16 2018-10-08 21:00:00+00:00                      1
17 2018-10-08 22:00:00+00:00                      1
18 2018-10-08 23:00:00+00:00                      1
19 2018-10-09 04:00:00+00:00                      4
20 2018-10-09 05:00:00+00:00                     20
21 2018-10-09 06:00:00+00:00                     39
22 2018-10-09 07:00:00+00:00                     51
23 2018-10-09 08:00:00+00:00                     51
24 2018-10-09 09:00:00+00:00                     52

Naturally, the DataFrame does only contain hours in which some user charged a vehicle. To ensure data consistency and have the data for **each** hour, we now want to add the missing entries and set "numberOfUsers" to 0 in these missing rows.

To simplify the following code, we have created a function to add the missing rows. The function first determines the time range of the given DataFrame. That is because when we previously took a look at the data, we were able to see that the data for the private site begins in October 2018, but the data for the public site begins in April 2018. We want to prevent that several months with entries of "0" are added to the private DataFrame because they have no insight for our model, whereas the 0 values in the actual time range of each site provide insight.

After this step, the function merges the two DataFrames and fills every missing value with 0. That leads to our missing rows being added and set to 0. Due to the nature of the number of users being no decimal number, we also convert the variable to an integer.

In [80]:
def add_missing_hours(df):
    # First create a DataFrame with all hours of the basis DataFrame
    all_hours = pd.DataFrame(pd.date_range(start=df['Hour'].min(), end=df['Hour'].max(), freq='H'), columns=['Hour'])
    
    # Then, merge the two DataFrame on the 'Hour' column and fill missing values with 0
    df = pd.merge(all_hours, df, on='Hour', how='left').fillna(0)
    
    return df

timeseries_private = add_missing_hours(timeseries_private)
timeseries_private['numberOfUsers_private'] = timeseries_private['numberOfUsers_private'].astype(int)

timeseries_public = add_missing_hours(timeseries_public)
timeseries_public['numberOfUsers_public'] = timeseries_public['numberOfUsers_public'].astype(int)

To provide additional context for our prediction, we now want to include a separate column that indicates if the hour is a weekday or not. 
Again, we use a function to change this for both DataFrames.

In [81]:
def add_weekday_column(df):
    df['isWeekday'] = df['Hour'].dt.weekday < 5
    return df

timeseries_private = add_weekday_column(timeseries_private)
timeseries_public = add_weekday_column(timeseries_public)

In [88]:
timeseries_public.head(200)
timeseries_private[timeseries_private['isWeekday'] == False]

Hour  numberOfUsers_private  isWeekday
115   2018-10-13 00:00:00+00:00                      0      False
116   2018-10-13 01:00:00+00:00                      0      False
117   2018-10-13 02:00:00+00:00                      0      False
118   2018-10-13 03:00:00+00:00                      0      False
119   2018-10-13 04:00:00+00:00                      0      False
...                         ...                    ...        ...
25694 2021-09-12 19:00:00+00:00                      0      False
25695 2021-09-12 20:00:00+00:00                      0      False
25696 2021-09-12 21:00:00+00:00                      0      False
25697 2021-09-12 22:00:00+00:00                      0      False
25698 2021-09-12 23:00:00+00:00                      0      False

[7344 rows x 3 columns]

The result proves that for example the 13th October 2018 was correctly labeled as no weekday since it was a Saturday. The new column works.